Added user dashboard

gsidebo opened the pull request:
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
  - [x] Mobile width screenshots
  - [x] Tag @ferdi or @pdpinch for review
- [x] Migrations
  - [x] Migration is backwards-compatible with current production code
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Closes #191 

#### What's this PR do?
Adds a user dashboard which shows program and course run enrollments

#### How should this be manually tested?
This gist will set you up with some program and course run enrollments: https://gist.github.com/gsidebo/5d2a5e9f9e4f9aa4508be59315bb404e

You can run it in a django shell/notebook.

#### Where should the reviewer start?
`courses/views.py` on back end, `DashboardPage.js` on front end

#### Any background context you want to provide?
- Design (expanded): https://projects.invisionapp.com/share/WDRNN8N2BT8#/screens/360105452
- Design (collapsed): https://projects.invisionapp.com/share/WDRNN8N2BT8#/screens/360105451
- This PR does not include the actual Open edX enrollment for a course. This just stores enrollment information on our end (actual enrollment issue: https://github.com/mitodl/mitxpro/issues/65)
- We decided to push the "Archived" section to a separate issue since it won't be needed for a while
- Some minor copy changes were made, and some elements from the design were intentionally dropped (course code, "Resume Course", redundant "Starts <date>" text)
- The `CourseRunEnrollment` model that existed in the `ecommerce` app wasn't really used for anything up until this point. After talking with George, we figured out that it wasn't necessary to link the enrollment model to a specific order and that it was sufficient just to link the user to the course run/program. I removed the old ecommerce model and defined the program and course run enrollment models in the `courses` app

#### Screenshots (if appropriate)
![ss 2019-05-09 at 16 58 45 ](https://user-images.githubusercontent.com/14932219/57486331-e6c93280-727b-11e9-8980-bc675d031966.png)

![ss 2019-05-09 at 16 58 38 ](https://user-images.githubusercontent.com/14932219/57486335-eaf55000-727b-11e9-991b-f92ee40ce4bc.png)

![ss 2019-05-09 at 16 49 03 ](https://user-images.githubusercontent.com/14932219/57486340-f0529a80-727b-11e9-97a6-a17de098b5be.png)

![ss 2019-05-09 at 16 48 15 ](https://user-images.githubusercontent.com/14932219/57486355-f779a880-727b-11e9-8d7d-2f047d5b5651.png)

![ss 2019-05-09 at 17 00 51 ](https://user-images.githubusercontent.com/14932219/57486378-05c7c480-727c-11e9-83cf-03a0d8c09c05.png)

gsidebo commented:
@pdpinch @Ferdi screenshots ^^^

pdpinch commented:
@gsidebo there's supposed to be a "flag" to distinguish programs and courses -- you can see it in InVision at https://projects.invisionapp.com/share/WDRNN8N2BT8#/screens/360105452

gsidebo commented:
@pdpinch Looks like an updated design. I'll implement that now

noisecapella commented:
@gsidebo Are you still working on the program/course run flag?

gsidebo commented:
@pdpinch @Ferdi @abdulkdawson Updated screenshots ^^^

gsidebo commented:
@noisecapella Just pushed

Ferdi commented:
Looks great to me

noisecapella commented:
The text "List of courses in this program:" is different from the mockup where the text is "The two courses in the certificate program are:". Does that matter here?

noisecapella commented:
The dropdown caret is narrow in this PR:
![Screenshot from 2019-05-09 17-12-52](https://user-images.githubusercontent.com/863262/57487119-ba161a80-727d-11e9-9e8a-f264162f5526.png)


 but it's wide in the mockup:

![Screenshot from 2019-05-09 17-12-30](https://user-images.githubusercontent.com/863262/57487112-b4b8d000-727d-11e9-84be-b9f8a9b8c635.png)

noisecapella commented:
The program flag is narrow and has a blurrier shadow in this PR:
![Screenshot from 2019-05-09 17-14-23](https://user-images.githubusercontent.com/863262/57487215-fe091f80-727d-11e9-97d7-153b2b0a56c7.png)

the mockup:
![Screenshot from 2019-05-09 17-14-23](https://user-images.githubusercontent.com/863262/57487236-0f522c00-727e-11e9-9396-5dc15dc605b8.png)

noisecapella commented:
The text for the title is a different color on this PR:
![Screenshot from 2019-05-09 17-16-59](https://user-images.githubusercontent.com/863262/57487351-56402180-727e-11e9-867d-763ac3b8f7fe.png)

the mockup:
![Screenshot from 2019-05-09 17-17-11](https://user-images.githubusercontent.com/863262/57487356-5b04d580-727e-11e9-99ec-944d0a416070.png)

Sorry if this is too nitpicky, feel free to punt on these issues if it makes sense to do so

noisecapella commented on mitxpro/utils.py:
```diff
@@ -122,3 +123,15 @@ def first_matching_item(iterable, predicate):
+    """
+    Partitions an iterable into two different iterables - the first does not match the given condition, and the second
+    does match the given condition.
+
```
Can you add an `Args:` here?

noisecapella commented on courses/api.py:
```diff
@@ -0,0 +1,52 @@
+            for program_enrollment in program_enrollments
+        )
+    )
+    program_course_ids = [course.id for course in program_courses]
```
We should make this a set if it's only used for lookup in the loop below

gsidebo commented:
@noisecapella 
1. Just changed the "View Courses" arrow to a material icon SVG
1. Changed the Courses and Programs header color
1. The "List of courses in this program:" copy change was by request
1. The ribbon width change was done to make it play nicer with course/program titles. I'm going to leave as-is. If we decide it needs to be changed, it can be done in a separate PR

gsidebo commented:
Just for posterity I'll mention that the actual process of enrolling a user in a course on Open edX is going to be done in https://github.com/mitodl/mitxpro/issues/65

noisecapella commented on ecommerce/factories.py:
```diff
@@ -1,4 +1,4 @@
-"""
+"" """
```
Typo?

noisecapella commented on mitxpro/utils_test.py:
```diff
@@ -62,3 +63,13 @@ def test_get_field_names():
     }
+
+
+def test_partition():
```
Can you add another test or parameterize this to test the default `bool` predictate?

noisecapella commented on static/js/components/Ribbon.js:
```diff
@@ -0,0 +1,40 @@
+import React from "react"
```
Can you add `// @flow` here?
